controller/MiNiFiController.cpp

```diff
@@ -0,0 +1,196 @@
+  std::string minifiHome;
+  if (const char *env_p = std::getenv(MINIFI_HOME_ENV_KEY)) {
+    minifiHome = env_p;
+    logger->log_info("Using MINIFI_HOME=%s from environment.", minifiHome);
+  } else {
+    logger->log_info("MINIFI_HOME is not set; determining based on environment.");
+    char *path = nullptr;
+    char full_path[PATH_MAX];
+    path = realpath(argv[0], full_path);
+
+    if (path != nullptr) {
+      std::string minifiHomePath(path);
+      if (minifiHomePath.find_last_of("/\\") != std::string::npos) {
+        minifiHomePath = minifiHomePath.substr(0, minifiHomePath.find_last_of("/\\"));  //Remove /minifi from path
+        minifiHome = minifiHomePath.substr(0, minifiHomePath.find_last_of("/\\"));    //Remove /bin from path
+      }
+    }
+
+    // attempt to use cwd as MINIFI_HOME
+    if (minifiHome.empty() || !validHome(minifiHome)) {
+      char cwd[PATH_MAX];
+      getcwd(cwd, PATH_MAX);
+      minifiHome = cwd;
+    }
+
+  }
+
+  if (!validHome(minifiHome)) {
+    logger->log_error("No valid MINIFI_HOME could be inferred. "
+                      "Please set MINIFI_HOME or run minifi from a valid location.");
+    return -1;
+  }
+
+  std::shared_ptr<minifi::Configure> configuration = std::make_shared<minifi::Configure>();
+  configuration->setHome(minifiHome);
+  configuration->loadConfigureFile(DEFAULT_NIFI_PROPERTIES_FILE);
+
+  std::shared_ptr<logging::LoggerProperties> log_properties = std::make_shared<logging::LoggerProperties>();
+  log_properties->setHome(minifiHome);
+  log_properties->loadConfigureFile(DEFAULT_LOG_PROPERTIES_FILE);
+  logging::LoggerConfiguration::getConfiguration().initialize(log_properties);
+
+  auto stream_factory_ = std::make_shared<minifi::io::StreamFactory>(configuration);
+
+  std::string host, port, caCert;
+
+  if (!configuration->get("controller.socket.host", host) || !configuration->get("controller.socket.port", port)) {
+    std::cout << "MiNiFi Controller is disabled" << std::endl;
+    exit(0);
+  }
+
+  cxxopts::Options options("MiNiFiController", "MiNiFi local agent controller");
+  options.positional_help("[optional args]").show_positional_help();
+
+  options.add_options()  //NOLINT
+  ("help", "Shows Help")  //NOLINT
+  ("stop", "Shuts down the provided component", cxxopts::value<std::vector<std::string>>())  //NOLINT
+  ("start", "Starts provided component", cxxopts::value<std::vector<std::string>>())  //NOLINT
+  ("l,list", "Provides a list of connections or processors", cxxopts::value<std::string>())  //NOLINT
+  ("c,clear", "Clears the associated connection queue", cxxopts::value<std::vector<std::string>>())  //NOLINT
+  ("getsize", "Reports the size of the associated connection queue", cxxopts::value<std::vector<std::string>>())  //NOLINT
+  ("updateflow", "Updates the flow of the agent using the provided flow file", cxxopts::value<std::string>())  //NOLINT
+  ("getfull", "Reports a list of full connections");
+
+  auto result = options.parse(argc, argv);
```

Comment:
would be nice to catch unsupported options and redirect to the help in lieu of the exception, e.g.
```
$ ./bin/minificontroller -h
[2018-01-infol 09:46:26.263] [controller] [info] MINIFI_HOME is not set; determining based on environment.
[2018-01-infol 09:46:26.264] [org::apache::nifi::minifi::Properties] [info] Using configuration file located at /Users/apiri/Development/code/apache/nifi-minifi-cpp/build/nifi-minifi-cpp-0.4.0/conf/minifi.properties
[2018-01-infol 09:46:26.264] [org::apache::nifi::minifi::Properties] [info] Using configuration file located at /Users/apiri/Development/code/apache/nifi-minifi-cpp/build/nifi-minifi-cpp-0.4.0/conf/minifi-log.properties
libc++abi.dylib: terminating with uncaught exception of type cxxopts::option_not_exists_exception: Option ‘h’ does not exist
[1]    56831 abort      ./bin/minificontroller -h
```